server/re/src/msi_genquery2.cpp

```diff
@@ -0,0 +1,315 @@
+
+/// Reads the value of a column from a row within a GenQuery2 resultset.
+///
+/// \warning This microservice is experimental and may change in the future.
+/// \warning This microservice is NOT thread-safe.
+///
+/// GenQuery2 handles are only available to the agent which produced them.
+///
+/// \param[in]     _handle       The GenQuery2 handle.
+/// \param[in]     _column_index The index of the column to read. The index must be passed as a string.
+/// \param[in,out] _column_value The variable to write the value of the column to.
+/// \param[in]     _rei          This parameter is special and should be ignored.
+///
+/// \return An integer.
+/// \retval  0 On success.
+/// \retval <0 On failure.
+///
+/// \since 4.3.2
+auto msi_genquery2_column(MsParam* _handle, MsParam* _column_index, MsParam* _column_value, RuleExecInfo* _rei) -> int
+{
+    log_msi::trace(__func__);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_column_index);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_column_value);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_rei);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->type);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_column_index->type);
+
+    IRODS_MSI_REQUIRE_TYPE(_handle->type, STR_MS_T);
+    IRODS_MSI_REQUIRE_TYPE(_column_index->type, STR_MS_T);
+
+    IRODS_MSI_REQUIRE_VALID_POINTER(_handle->inOutStruct);
+    IRODS_MSI_REQUIRE_VALID_POINTER(_column_index->inOutStruct);
+
+    try {
+        const auto ctx_handle_index = std::stoll(static_cast<char*>(_handle->inOutStruct));
+
+        if (ctx_handle_index < 0) {
+            log_msi::error("{}: Unknown context handle.", __func__);
+            return SYS_INVALID_INPUT_PARAM;
+        }
+
+        auto iter = gq2_context.find(ctx_handle_index);
+
+        if (std::end(gq2_context) == iter) {
+            log_msi::error("{}: Unknown context handle.", __func__);
+        }
+
+        auto& ctx = iter->second;
+        const auto column_index = std::stoll(static_cast<char*>(_column_index->inOutStruct));
+
+        const auto& value = ctx.rows.at(ctx.current_row).at(column_index).get_ref<const std::string&>();
+        log_msi::debug("{}: Column value = [{}]", __func__, value);
+        fillStrInMsParam(_column_value, value.c_str());
+
+        return 0;
+    }
+    catch (const std::exception& e) {
+        log_msi::error("{}: {}", __func__, e.what());
+        return SYS_LIBRARY_ERROR;
+    }
+} // msi_genquery2_column
+
+/// Frees all resources associated with a GenQuery2 handle.
+///
+/// \warning This microservice is experimental and may change in the future.
+/// \warning This microservice is NOT thread-safe.
+///
+/// GenQuery2 handles are only available to the agent which produced them.
+///
+/// Users are expected to call this microservice when a GenQuery2 resultset is no longer needed.
+/// Failing to follow this rule can result in memory leaks. However, the handle and the resultset
+/// it is mapped to will be free'd on agent teardown.
```

Comment:
Should this be put under a `\warning` as well? Feels important.

Reply:
Sure. Making that a warning does seem better.